### PR TITLE
Add CentOS based DotNet 2.0 stack to default assembly

### PR DESCRIPTION
**Pull request by dharmit**
Signed-off-by: Dharmit Shah <dshah@redhat.com>

### What does this PR do?
Adds CentOS based dotnet 2.0 container image to the assembly

### What issues does this PR fix or reference?
#125

### CODE REVIEW

**Comment by dharmit** [comment]
cc: @l0rd

File: recipes/centos_dotnet20/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2012-2017 Red Hat, Inc
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
+
+FROM registry.centos.org/che-stacks/centos-jdk8:latest
```
Comment (l0rd) [inline recipes/centos_dotnet20/Dockerfile]: @dharmit shouldn't we use image `centos-stack-base`  as base stack? The base stack has everything is needed to run on Che (jdk included). `centos-jdk8` adds maven and tomcat on top of the base stack. Do we need them for .NET?

Comment (dharmit) [reply recipes/centos_dotnet20/Dockerfile]: Makes sense. I'll change it.

**Comment by dharmit** [comment]
@l0rd @omajid you can already pull the image and take it for a spin. I'll be adding stacks entry right after this is merged. `docker pull registry.centos.org/che-stacks/centos-dotnet20`

**Comment by omajid** [comment]
Adding to path is necessary, but not sufficient. Please see https://github.com/eclipse/che-dockerfiles/issues/125#issuecomment-334454846. I would recommend sourcing `scl_enable` *somehow*.

**Comment by eivantsov** [comment]
The recipe itself is ok, but we gotta solve https://github.com/eclipse/che-dockerfiles/issues/125 first

**Comment by l0rd** [comment]
@dharmit @omajid I would like to merge this PR even if omnisharp doesn't work with this stack. It would provide a first basic .NET support on openshift.io (build tools + syntax highlighting). We can then iterate with a new version of this image as soon as the language server work with our .NET core distribution.

But there are a couple of things that should be addressed to make this PR mergeable:

1. The `ENTRYPOINT` defined in this Dockerfile overrides the `ENTRYPOINT` of the [base centOS stack](https://github.com/eclipse/che-dockerfiles/blob/master/recipes/stack-base/centos/Dockerfile#L61). As a side effect the original script [entrypoint.sh](https://github.com/eclipse/che-dockerfiles/blob/master/recipes/stack-base/centos/entrypoint.sh) (that is similar official rhel image [generate_container_user](https://github.com/redhat-developer/s2i-dotnetcore/blob/master/2.0/runtime/contrib/etc/generate_container_user)) is never called and makes the image unusable in OpenShift.

2. Che [commands](https://github.com/eclipse/che/blob/master/ide/che-core-ide-stacks/src/main/resources/stacks.json#L465) are run using `docker exec` (not `docker run`) and changing the entrypoint doesn't have the desired effect. For example:

```
$ docker run -d --name centos_dotnet20 \
                 registry.centos.org/che-stacks/centos-dotnet20 \
                 tail -f /dev/null
05521aef50a6bd355160c2dcaf8876f71231a88b5d0b2c0d6ea5e91e4aeb8e7e
$ docker exec -ti centos_dotnet20 dotnet --version
oci runtime error: exec failed: container_linux.go:265: starting container process caused "exec: \"dotnet\": executable file not found in $PATH"
```
 
To address the first point I would recommend not overwriting the `ENTRYPOINT`.  You could update the base stack [entrypoint.sh](https://github.com/eclipse/che-dockerfiles/blob/master/recipes/stack-base/centos/entrypoint.sh) script instead but I can't see any real benefit (`docker run centos_dotnet20 dotnet ...` is not used in Che)

I can't see any good solution to address the second point. And I would continue to use `scl enable rh-dotnet20 'dotnet ...'` for commands until we have found it.

**Comment by omajid** [comment]
Thanks for looking into this!

For `ENTRYPOINT`, the goal is to invoke a bash script, which will end up sourcing `PROMPT_COMMAND`. Looks like the base entrypoint does that already (except unsetting the env var) so that shouldnt be a problem.

For the second point, if there is no way to manipulate what `exec` will invoke, then `scl enable rh-dotnet20 -- dotnet $ACTUAL_VERB_HERE` seems like the simplest way.

**Comment by eivantsov** [comment]
With `dotnet` being in `$PATH` it's also possible to add envs to the Dockerfile directly - whatever is exported in the enable file

**Comment by dharmit** [comment]
In that case, it seems to me that the original PR which was similar to this [Dockerfile](https://github.com/dharmit/dockerfiles/blob/master/dotnet20/Dockerfile) is good. I need to change the `FROM` in that file. And instead of using `scl_enable` script, copy the contents of `/opt/rh/rh-dotnet20/enable` to `~/.bashrc` like we did earlier. 

@l0rd @omajid @eivantsov Does that sound OK?

**Comment by eivantsov** [comment]
If dotnet isn't in $PATH, languagesevrer won't work. But in case of this image it won't work because of the dotnet itself.

**Comment by l0rd** [comment]
@dharmit it's ok for me because it's consistent with other stacks. The `scl` problem for `docker exec` and `docker run` is common for all the centOS stack and we have an [open issue](https://github.com/redhat-developer/rh-che/issues/371) that is planned for this sprint.

**Comment by l0rd** [comment]
@dharmit are you able to update the PR or do you want us to work on it. I would like to have it merged today.

**Comment by l0rd** [comment]
I have created #144 to replace this PR

**Comment by l0rd** [comment]
The CentOS dotNet 2.0 stack has been added with #144. Closing this one.

**Pull request closed without merging**